### PR TITLE
Implementing a complete CLI support for the web-api

### LINKED ISSUE #6: install dependencies should get listed in the setup.py
after `pip install pyamplipi` one can not just start using the library without also including its dependencies: 

even a simple `from pyamplipi.amplipi import AmpliPi` yields an error (in that case: missing dependency aiohttp)

most straightforward approach would be to have the setup.py take over the dependencies from the requirements.txt file (or using a tool like poetry to handle it)

### LINKED ISSUE #7: Announcement model requires more fields than the API does.
The model for `Announcement` as defined at https://github.com/brianhealey/pyamplipi/blob/main/pyamplipi/models.py#L168

requires fields `media`, `vol_f` and `source_id`

while the API documentation at http://api.amplipi.com/#tag--announce only requires the media field and applies sensible defaults for the others

I can provide a PR 

However - am a bit confused on the `vol_f: float`  versus `vol: int` usage?
The API documentation does only mention the `vol: int` variant (as well as a `vol_delta: int` on zones)  Its semantics and use are not quite clear to me.

### LINKED ISSUE #15: introduce flake8 lint checking to help avoid programming errors and consistent styling
Applying flake8 on the current codebase gives a nice report:

```
$ flake8 . --count --exit-zero --max-complexity=10 --max-line-length=132 --statistics --exclude venv,.venv,.eggs,.tox,.git,__pycache__,*.pyc
./pyamplipi/amplipi.py:1:1: F401 'json' imported but unused
./pyamplipi/amplipi.py:110:44: F541 f-string is missing placeholders
./pyamplipi/cli.py:9:1: F401 'pyamplipi.models.ZoneUpdate' imported but unused
./pyamplipi/cli.py:9:1: F401 'pyamplipi.models.ZoneUpdateWithId' imported but unused
./pyamplipi/models.py:35:1: E302 expected 2 blank lines, found 1
./setup.py:22:9: F841 local variable 'reqs' is assigned to but never used
1     E302 expected 2 blank lines, found 1
3     F401 'json' imported but unused
1     F541 f-string is missing placeholders
1     F841 local variable 'reqs' is assigned to but never used
6
```

Much of this is strictly "style conformity" - but the last one actually helps discovering a recently introduced bug.

If you guys think it useful I would like to prepare a PR that 
- adds flake8 to `requirements-dev.txt` + update `setup.py` to be aware of it (as already hinted at in #6) 
- fixes the current remarked issues in the code 
- introduce a github-ci-workflow to run the flake8 (preferably via `make lint` goverened by introducing a `Makefile`)

### LINKED ISSUE #16: feature request -- cli support for all functions in the API
The current `cli.py` functions as a nice introduction/example on how to use the library for very specific scripted automation needs.  Actually `example.py` would probably be a better name, since it does not provide a full CLI coverage of all the functions provided by the library. 

We could however introduce such coverage via a `__main__.py`  

It should:
* be hooked up in the setup.py as a so called console-script 'entry-point'  via `setup( entry_points={'console_scripts'= ['pyamplipi = pyamplipi.__main__:main']})`
* read recurring settings like `AMPLIPI_API_URL` and `AMPLIPI_TIMEOUT` from the environment through `.env` + [`python-dotenv`](https://pypi.org/project/python-dotenv/)
* allow to override those on the command line via `-a «api_url»`  and `-t «timeout»`
* apply configurable logging
* provide consistent error-codes if things fail
* use [`argparse`](https://pypi.org/project/argparse/) to cover all of the api via some smart design of the args: 
```
pyamplipi [-a api_url] [-t timeout] TOPIC CMD ...

pyamplipi status list  # produces comprehensible output listing / overview
pyamplipi status get > output.json  
pyamplipi status set < input.json

pyamplipi source list  # produces comprehensible output listing / overview
pyamplipi source get «sourceid» | *  > output.json  # * wildcard for getting all
pyamplipi source set «sourceid» < update.json

pyamplipi zone list  # produces comprehensible output listing / overview
pyamplipi zone get «zoneid» | * > output.json
pyamplipi zone set «zoneid» < update.json

pyamplipi group list  # produces comprehensible output listing / overview
pyamplipi group get «groupid» | * > output.json
pyamplipi group set «groupid» < update.json
pyamplipi group new < input.json  # returns created groupid
pyamplipi group del «groupid» 

pyamplipi stream list  # produces comprehensible output listing / overview
pyamplipi stream get «streamid» | * > output.json
pyamplipi stream set «streamid» < update.json
pyamplipi stream new < input.json  # returns created groupid
pyamplipi stream del «groupid» 
pyamplipi stream play «groupid»  
pyamplipi stream pause «groupid»  
pyamplipi stream stop «groupid»  
pyamplipi stream next «groupid»  
pyamplipi stream prev «groupid»  

pyamplipi announce «media_url» [«vol_f»]  # plays the announcement

pyamplipi preset list  # produces comprehensible output listing / overview
pyamplipi preset get «presetid» | * > output.json
pyamplipi preset set «presetid» < update.json
pyamplipi preset load  «presetid» < input.json
pyamplipi preset new < input.json  # returns created presetid
pyamplipi preset del «presetid» 
```

note: 
* edited the above after starting on this --> reordering CMD and TOPIC arguments turned out to implement more natural
* not all preset related endpoints are available in the library yet (might be a recent v0.1.9 ? addition) -- so those would need to be added too
* have to figure out the difference between set-preset `PATCH /api/presets/{pid}` and load-preset `POST /api/presets/{pid}/load` and how the CLI might maybe either hide or support both in the most natural way

I would be glad to provide a PR introducing the above as I see it a logical fit to what the lib is offering.

But I would also understand if you consider this as 'out of scope' and tell me to buzz off and implement this in a separate project.

### PR DESCRIPTION
**Pull request by marc-portier**
fixes a bunch of outstanding things:
* closes #6 (already closed but with a missing return)
* closes #7 (as far more extra model changes were applied to match up with v0.1.9)
* closes #15 (as flake8 was useful in safeguarding this work as well)
* closes #16 (the main purpose)
* starts on #12 with minimal `make check`

### CODE REVIEW

**Comment by linknum23** [comment]
Let me know when this is ready and I'll review it!

**Comment by marc-portier** [comment]
Please do review and consider the merge already
- I consider it "ready enough" for that  (specially after resolving the homeassistant dependency conflicts this morning) 
- Note that the last/late added commit also bumps the version to 0.4.10-dev indicating the difference with the latest stable 0.4.9
- also because I was planning to move my attention slightly away from this for the first days / weeks now - leaving others time to review / comment / tune
- Even if this still contains some minor "todo" remarks that I left around --> in fact: if you prefer (for ease of tracking) I can create separate issues for those? Just say the word.
- and will gladly update the PR I created for #14 after merging this


Which makes me think about next steps on this library:
after this rather big merge (which turns out more important because of the 0.1.9 update than because of the cli support IMHO) I would like is us to 
(1) discuss the testing (#12) approach and 
(2) decide on some kind of general / long-term release plan (i.e. how would we want to have the lib follow/sync up with the amplipi versions) 
(3) list which of the open minor stuff we want to include in the short-term next release?


BTW: the above does mean I want to be involved and active in collaborating on those goals.
But I also understand if my current  planning, eagerness and availability doesn't fit your schedule - conversation to be had should make details like that clear?


Finally, my personal plan is now to start applying all of this stuff into my personal setup / local home development first.
I assume that work might show other bugs / feature requests from the real-life testing it will involve.
Also: I was planning to share my design goals and plans on the community discord one of these days - that might lead to consider "higher level" functionality to be welcome as part of this lib rather than in my personal repo only?

**Comment by linknum23** [comment]
I'm just starting to play with the cli after making some changes to make it work in 3.8. This is fantastic! The output looks great.
What python version are you developing in? I noticed the setup.py still has 3.5/3.6 listed but I'm assuming this is being developed in 3.9-3.11?

**Comment by marc-portier** [comment]
well, glad you like it (the nice tables come - almost for free - from the [tabulate](https://pypi.org/project/tabulate/) package.)

I am on ubuntu 22.04 which comes with python 3.10.6
am eager to learn how I can easily dev/test for multiple py versions (maybe also something the CI approach can help with?)

File: .github/workflow/python-app.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python
+
+name: Python application
```
Comment (linknum23) [inline .github/workflow/python-app.yml:3]: We will likely need brian to enable these and in the future move to micro-nova's github so we don't use up his minutes.

File: pyamplipi/docs/cli.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
+# Commane Line Interface
```
Comment (linknum23) [inline pyamplipi/docs/cli.md:1]: We probably will want to link to this from the main README

File: pyamplipi/docs/cli.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+Full listing and more explanation are offered by passing `-h` or `--help` either in the general or the specific switches.
+
+Note that all switches have a `--Long` and a `-L` (short) variant
+
+## Getting stdout Output
+
+The `get` `COMMAND` generally will produce a json formatted output representation on stdout.
+
+This allows to pipe the output to json handling tools (like `jq`) or simply redirect the output to a file using `> output-here.json`
```
Comment (linknum23) [inline pyamplipi/docs/cli.md:45]: `jq` is pretty great for simple stuff

**Review by linknum23 — COMMENTED** [review]
We've been playing around with the interface for a couple of hours and everything is very usable.

I found a couple of small issues in shell mode that kicked me out:

Entering the topic correctly but forgetting the action:
```
ampsh > zone 		
usage: pyamplipi zones [-h] ACTION ...
pyamplipi zones: error: the following arguments are required: ACTION```
```
 I also get kicked out when trying to print help:

```
ampsh > -h  
usage: pyamplipi [-h] [-l LOGCONF_FILE.yml] [-a API_URL] [-t SECONDS] TOPIC ...

CLI for interactive amplipi /api calls

options:
  -h, --help            show this help message and exit
  -l LOGCONF_FILE.yml, --logconf LOGCONF_FILE.yml
...
```

Also when running this through my standard typechecker setup of pylint + mypy I found and fixed some type issues I added them to the PR in the hopes that they are helpful. Feel free to remove them.

File: pyamplipi/docs/cli.md (hunk 2 above)
Comment (marc-portier) [reply pyamplipi/docs/cli.md:1]: > We've been playing around with the interface for a couple of hours and everything is very usable.
> 
> I found a couple of small issues in shell mode that kicked me out:
> 
> Entering the topic correctly but forgetting the action:
> 

yes, and this moves the argparse to go into `--help` mode which in turn does a system exit
I tried resolving that through `exit_on_error = False`  but only helped for parse-errors that don´t trigger the print_help()

haven´t found another solution yet

> 
> Also when running this through my standard typechecker setup of pylint + mypy I found and fixed some type issues I added them to the PR in the hopes that they are helpful. Feel free to remove them.

interesting.

if that typechecking is available as a standalone python package than we should look into  (1) adding it to the `requirements-dev.txt` and (2) include in the `make check` 


I have no intention at all to remove your additions / modifications -- will investigate later, but I am all ok with you proceeding with the merge...

**Comment by linknum23** [comment]
On stack overflow I saw some mention of a way to hook into argparses
on_exit and on_error handling. I'll see if I can find it again.

As for the type checker I'll add it. It's just mypy with some settings I
think?

**Comment by linknum23** [comment]
I added the mypy checks, and fixed some of the problems it identified. I'll merge this in so we can keep the momentum on this!

**Pull request merged**